### PR TITLE
[FLINK-19624][table-planner-blink] Update deadlock break-up algorithm to cover more cases

### PR DESCRIPTION
**Pull request by TsReaper**
## What is the purpose of the change

Current deadlock breakup algorithm in Blink planner fails to cover some cases. See FLINK-19624 and the appendix of the [design doc](https://docs.google.com/document/d/1qKVohV12qn-bM51cBZ8Hcgp31ntwClxjoiNBUOqVHsI) for more information.

We're going to introduce a new deadlock breakup algorithm to solve this problem. This algorithm is also used in the upcoming multiple input node creation algorithm.

## Brief change log

 - Introduce a new deadlock breakup algorithm.

## Verifying this change

This change added tests and can be verified as follows: Run the added test cases.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/reuse/InputPriorityConflictResolver.java [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+ * <a href="https://docs.google.com/document/d/1qKVohV12qn-bM51cBZ8Hcgp31ntwClxjoiNBUOqVHsI">design doc</a>.
+ */
+@Internal
+public class InputPriorityConflictResolver {
+
+	private final List<ExecNode<?, ?>> sinks;
+
+	private TopologyGraph graph;
+
+	public InputPriorityConflictResolver(List<ExecNode<?, ?>> sinks) {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/reuse/InputPriorityConflictResolver.java:103]: nit: an `ExecNode` may be not `BatchExecSink` here, use roots ?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/reuse/InputPriorityConflictResolver.java [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+				if (!hasAncestor) {
+					ret.add(node);
+				}
+			}
+		};
+		node.accept(ancestorVisitor);
+		return ret;
+	}
+
+	private BatchExecExchange createExchange(ExecNode<?, ?> node, int idx) {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/reuse/InputPriorityConflictResolver.java:208]: could you provide more test cases to cover different branches ?

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/plan/reuse/InputPriorityConflictResolverTest.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+	public void testDetectAndResolve() {
+		// 0 --------(P0)----> 1 --(P0)-----------> 7
+		//  \                    \-(P0)-> 2 -(P0)--/
+		//   \-------(P0)----> 3 --(P1)-----------/
+		//    \------(P0)----> 4 --(P10)---------/
+		//     \              /                 /
+		//      \    8 -(P0)-<                 /
+		//       \            \               /
+		//        \--(E0)----> 5 --(P10)-----/
+		// 6 ---------(P100)----------------/
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/plan/reuse/InputPriorityConflictResolverTest.java:150]: add some comments to explain `Pxx`, `Exx`

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/nodes/exec/TestingExecNode.scala [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+import org.apache.calcite.rel.metadata.{Metadata, RelMetadataQuery}
+import org.apache.calcite.rex.{RexNode, RexShuttle}
+import org.apache.calcite.util.{ImmutableBitSet, Litmus}
+
+import java.util
+
+/**
+ * [[BatchExecNode]] for testing purpose.
+ */
+class TestingExecNode extends BatchExecNode[BatchPlanner] with BatchPhysicalRel {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/nodes/exec/TestingExecNode.scala:41]: rename to `TestingBatchExecNode` ?
also extends from `AbstractRelNode`, which has implemented many methods.

**Review by godfreyhe — COMMENTED** [review]
Thanks for the contribute @TsReaper, LGTM overall, I left a few comments

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/reuse/InputPriorityConflictResolver.java (hunk 2 above)
Comment (TsReaper) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/reuse/InputPriorityConflictResolver.java:208]: Existing tests in `DeadlockBreakupTest` already covers the possible branches.

**Comment by godfreyhe** [comment]
@flinkbot run azure

**Review by godfreyhe — APPROVED** [review]
LGTM, will merge it once the tests pass

**Pull request merged**